src/audio/src/src.c

```diff
@@ -503,30 +502,58 @@ static int src_stream_pcm_source_rate_check(struct ipc4_config_src cfg,
 	return 0;
 }
 
+static int src_stream_pcm_sink_rate_check(struct ipc4_config_src cfg,
+					  struct sof_ipc_stream_params *params)
+{
+	if (cfg.sink_rate && params->rate != cfg.sink_rate)
+		return -EINVAL;
+
+	return 0;
+}
+
 /* In ipc4 case param is figured out by module config so we need to first
  * set up param then verify param. BTW for IPC3 path, the param is sent by
  * host driver.
  */
-static void src_set_params(struct comp_dev *dev, struct sof_ipc_stream_params *params)
+static int src_set_params(struct processing_module *mod)
 {
-	struct processing_module *mod = comp_get_drvdata(dev);
+	struct sof_ipc_stream_params src_params;
+	struct sof_ipc_stream_params *params = mod->stream_params;
 	struct comp_data *cd = module_get_private_data(mod);
+	enum sof_ipc_frame frame_fmt, valid_fmt;
+	struct comp_dev *dev = mod->dev;
 	struct comp_buffer *sinkb;
 	struct comp_buffer __sparse_cache *sink_c;
+	int ret;
 
-	memset(params, 0, sizeof(*params));
-	params->channels = cd->ipc_config.base.audio_fmt.channels_count;
-	params->rate = cd->ipc_config.base.audio_fmt.sampling_frequency;
-	params->sample_container_bytes = cd->ipc_config.base.audio_fmt.depth / 8;
-	params->sample_valid_bytes = cd->ipc_config.base.audio_fmt.valid_bit_depth / 8;
-	params->frame_fmt = dev->ipc_config.frame_fmt;
-	params->buffer_fmt = cd->ipc_config.base.audio_fmt.interleaving_style;
-	params->buffer.size = cd->ipc_config.base.ibs;
+	src_params = *params;
+	src_params.channels = mod->priv.cfg.base_cfg.audio_fmt.channels_count;
+	src_params.buffer_fmt = mod->priv.cfg.base_cfg.audio_fmt.interleaving_style;
+	src_params.rate = cd->ipc_config.sink_rate;
+
+	/* Get frame_fmt and valid_fmt */
+	audio_stream_fmt_conversion(mod->priv.cfg.base_cfg.audio_fmt.depth,
+				    mod->priv.cfg.base_cfg.audio_fmt.valid_bit_depth,
+				    &frame_fmt, &valid_fmt,
+				    mod->priv.cfg.base_cfg.audio_fmt.s_type);
+
+	src_params.frame_fmt = frame_fmt;
+	component_set_nearest_period_frames(dev, src_params.rate);
 
 	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
 	sink_c = buffer_acquire(sinkb);
-	sink_c->stream.rate = cd->ipc_config.sink_rate;
+	ret = buffer_set_params(sink_c, &src_params, true);
 	buffer_release(sink_c);
+
+	/* To pass source and sink rate check functions, alternatively skip the
+	 * checks for IPC4.
+	 */
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
+		params->rate = cd->ipc_config.base.audio_fmt.sampling_frequency;
+	else
+		params->rate = cd->ipc_config.sink_rate;
```

Comment:
@singalsu I think we should use params->rate = cd->ipc_config.sink_rate; for both directions